### PR TITLE
Timestamp extraction and automatic chunking

### LINKED ISSUE #3: [Feature] Word level timestamps
### Feature Description

Hello, 
is there any way to get word level timestamps or is there a plan to add this capability to the model ?
Thanks

### Use Case

_No response_

### PR DESCRIPTION
**Pull request by urroxyz**
This PR introduces two requested features to the `ASRInferencePipeline`: **accurate timestamp extraction** (word/character level) and **automatic audio chunking** for long-form files. Closes #3.

### Summary of changes
- `transcribe()` now returns a tuple `(transcripts, timestamps)`.
    - Uses frame boundary analysis from the emission logits for CTC models.
    - Uses a novel Dynamic Time Warping (DTW) approach on the decoder's Cross-Attention maps (Text Query $\to$ Audio Key) to enforce monotonic alignment and prevent attention collapse for LLM-based models.
    - Includes language-agnostic logic to automatically detect non-spaced scripts (CJK, Thai, Lao, Khmer, Myanmar) and switch from word-level to character-level timestamps.
- Input audio exceeding the model's limit of 40s is now automatically chunked. The `chunk_len` parameter allows users to define this window (default `None` preserves strict behavior, setting it enables chunking).
- All implemented using pure `torch` and `numpy` without adding heavy dependencies like `scipy` or `librosa`.

## Why?
This addresses an important utility gap compared to similar libraries like OpenAI's Whisper. Community demand for timestamp alignment and long-form transcription has been high.

Relevant @mentions: @ijean, @artemru, @d-cota, @huanglizhuo, @wikioai, @nikopartanen, @Aunali321, @marcelgoya, and some others.

## How?
- Created a new module to handle alignment strategies.
    - Maps greedy decoding paths to time frames and interpolates text units between non-blank token transitions for CTC.
    - Performs a secondary "teacher-forced" forward pass with hooks on `StandardMultiheadAttention` for LLMs. It captures QK similarity matrices, aggregates them across layers, and applies a custom NumPy-based DTW algorithm to find the optimal alignment path.
- Updated `transcribe` to handle waveform loading and splitting based on `chunk_len`.
    - It processes chunks sequentially, aligns them individually, and recalculates global time offsets before returning the merged result.

## Test plan
Verified on both CTC and LLM architectures using English and Japanese.

### 1. CTC model for word-level alignment
```python
from omnilingual_asr.models.inference.pipeline import ASRInferencePipeline

# Load CTC model
pipeline = ASRInferencePipeline(model_card="omniASR_CTC_300M")

audio_files = ["IMDA_conversation.wav"] # English audio example
transcriptions, timestamps = pipeline.transcribe(audio_files, chunk_len=40, batch_size=2) # Request chunking at 40s intervals

print(transcriptions)
print(timestamps)
```

**Output:**
```text
['and then we will need some documents from you...']
[[{'word': 'and', 'start': 8.21, 'end': 8.44}, {'word': 'then', 'start': 8.42, 'end': 8.77}, ...(TRUNCATED)...]]
```

### 2. LLM model and script detection
```python
from omnilingual_asr.models.inference.pipeline import ASRInferencePipeline

# Load LLM model
pipeline = ASRInferencePipeline(model_card="omniASR_LLM_300M")

audio_files = ["ado.wav"] # Japanese audio example
transcriptions, timestamps = pipeline.transcribe(audio_files, chunk_len=40, batch_size=2)

print(transcriptions)
print(timestamps)
```

**Output:**
```text
['まずどんな勉強してたか...']
[[{'char': 'ま', 'start': 0.0, 'end': 0.32}, {'char': 'ず', 'start': 0.32, 'end': 0.42}, {'char': 'ど', 'start': 0.42, 'end': 0.54}, ...(TRUNCATED)...]] # Accurate character-level timestamps for CJK scripts via LLM attention
```

### CODE REVIEW

**Comment by urroxyz** [comment]
@artemru

**Comment by seohyunjun** [comment]
@urroxyz 
Is it possible to receive the output in segment units as well?
The word-level chunks are too detailed to use as subtitles.
Thank you for developing such good code.

**Comment by seohyunjun** [comment]
@urroxyz 

It seems that, unlike Whisper, this method doesn’t use time tokens to store the segment’s end time.

As you mentioned, it looks like I could infer silent sections with a VAD filter to separate segments, or handle sentence endings with a rule-based approach. Thank you for the idea. I’ve used a diarization model before to split segments, but I’ll need to test whether it will be effective in this case as well.

Even with just the code you provided, it has already been extremely helpful for the project I’m working on.

Thank you!!

**Comment by urroxyz** [comment]
@seohyunjun Yes, it is unfortunate that the OmniASR models lack punctuation and native "time tokens" for boundary detection. I'm trying to see if there is a low-level hack to reconcile these things, but it's unlikely I'll discover anything reliable.

Glad that it's still a help, though. Let me know if you need any additional information or support for your project, I'm a student of computational linguistics and love all things ASR and subtitling!

**Comment by Fannovel16** [comment]
@urroxyz Is it possible to do forced alignment (get timestamp from ground truth transcription) with your approachs?

**Comment by d-cota** [comment]
Any ETA regarding the merge?

File: src/omnilingual_asr/models/inference/align.py [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+    waveform: Tensor,
+    transcript: str,
+    lang: Optional[str] = None
+) -> List[dict]:
+
+    if not transcript.strip():
+        return []
+
+    # Narrow type for MyPy
+    model = cast(Wav2Vec2LlamaModel, pipeline.model)
```
[the remark this answers is not in the document]
Comment (urroxyz) [reply src/omnilingual_asr/models/inference/align.py]: It should now be fixed with the latest commit.

**Comment by cirquit** [comment]
Hi @urroxyz, thanks for this contribution! From a first glance, this looks nice - I've allowed the lint&test gh action, please try to fix the errors first until I find the time to fully review your PR.

**Comment by seohyunjun** [comment]
@urroxy
check this PR plz.
Apply ruff format
https://github.com/urroxyz/omnilingual-asr/pull/1

**Comment by rocety27** [comment]
Any update on this? @cirquit ?

**Comment by cirquit** [comment]
> Any update on this? @cirquit ?

Sorry for the silence - this is still on my radar. I sadly can't provide a timeline for a review given work priorities.

**Review by Fannovel16 — APPROVED** [review]

**Comment by urroxyz** [comment]
Thank you for the approval, @Fannovel16. I assume the library has had lots of updates in the past few months. Does the code still integrate properly with the codebase?

Also, I changed the title because it was a little too quirky. Lol.

**Comment by rocety27** [comment]
@urroxyz This might not integrate seamlessly with unlimited variants I think
